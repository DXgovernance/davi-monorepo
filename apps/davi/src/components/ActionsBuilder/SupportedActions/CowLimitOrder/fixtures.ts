import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber, utils } from 'ethers';
import { MOCK_ADDRESS } from 'hooks/Guilds/ens/fixtures';
import CowSettlement from 'contracts/CowSettlement.json';

const CowSettlementContract = new utils.Interface(CowSettlement.abi);

export const limitOrderDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.COW_SWAP_LIMIT_ORDER,
  function: CowSettlementContract.getFunction('setPreSignature'),
  to: '0xfFb1cd0F95368DDd06D556161c5D3d9f0f4Fe6d2',
  value: BigNumber.from(0),
  args: {
    orderUid: '',
    signed: '',
  },
  optionalProps: {
    sellAmount: BigNumber.from('5000000000000000000'),
    buyToken: MOCK_ADDRESS,
    sellToken: MOCK_ADDRESS,
    functionSignature: '',
  },
};
