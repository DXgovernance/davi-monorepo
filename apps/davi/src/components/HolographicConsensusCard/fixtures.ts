import { BigNumber } from 'ethers';
import { IStakes } from './types';

export const mockProposalStakeDetails: IStakes = {
  for: [
    {
      id: '1',
      staker: '0xaf8eb8c3a5d9d900aa0b98e3df0bcc17d3c5f698',
      amount: BigNumber.from('100000000000000000000'),
    },
    {
      id: '2',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('100000000000000000000'),
    },
    {
      id: '3',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('336000000000000000000'),
    },
    {
      id: '4',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('236600000000000000000'),
    },
    {
      id: '5',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('47138000000000000000'),
    },
  ],
  against: [
    {
      id: '6',
      staker: '0x84eeb305da0a4309a696d43de9f79f04e66eb4f8',
      amount: BigNumber.from('100000000000000000000'),
    },
  ],
};

export const mockProposalStakeDetailsEmpty = {
  for: [],
  against: [],
};
