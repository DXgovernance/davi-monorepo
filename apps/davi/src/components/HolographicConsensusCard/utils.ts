import { BigNumber, FixedNumber } from 'ethers';
import { Proposal } from 'types/types.guilds.d';
import { IStakes } from './types';

export const calculateSpeedometerValue = (
  totalStaked: Proposal['totalStaked']
): number => {
  if (!totalStaked) return 0;

  const stakedAgainst = totalStaked[0];
  const stakedFor = totalStaked[1];

  const stakeDifference = stakedFor.sub(stakedAgainst);
  let percentage: number;

  if (stakeDifference.gt(0)) {
    percentage = stakeDifference.mul(100).div(stakedFor).toNumber();
  } else if (stakeDifference.eq(0)) {
    percentage = 0;
  } else {
    percentage = stakeDifference.mul(100).div(stakedAgainst).toNumber();
  }
  return 5000 + percentage * 50;
};

export const calculateBNPercentage = (
  value: BigNumber,
  percentage: string | number
) => {
  // percentage should be an integer (before being divided by 100)
  // if a percentage has decimals, it trims them
  // TODO: handle decimals?
  if (typeof percentage === 'string') {
    if (percentage.includes('.')) {
      const percentageSplit = percentage.split('.');
      percentage = percentageSplit[0];
    }
  }
  if (typeof percentage === 'number') {
    if (percentage % 1 !== 0) {
      percentage = Math.trunc(percentage);
    }
  }

  const valueFixed = FixedNumber.fromValue(value);
  const percentageFixed = FixedNumber.fromValue(BigNumber.from(percentage));
  const hundredFixed = FixedNumber.fromValue(BigNumber.from(100));

  const resultFixed = valueFixed
    .mulUnsafe(percentageFixed)
    .divUnsafe(hundredFixed);

  const resultFixedInteger = resultFixed.toString().split('.')[0];
  const resultBN = BigNumber.from(resultFixedInteger);
  return resultBN;
};

export const checkUserStakeOption = (
  userAddress: string,
  stakeDetails: IStakes
) => {
  const userStakeFor = stakeDetails.for.find(
    stake => stake.staker.toLowerCase() === userAddress.toLowerCase()
  );
  if (userStakeFor) return 'for';

  const userStakeAgainst = stakeDetails.against.find(
    stake => stake.staker.toLowerCase() === userAddress.toLowerCase()
  );
  if (userStakeAgainst) return 'against';

  return null;
};

export const calculatePotentialReward = (
  stakes: IStakes,
  userCurrentStake: BigNumber,
  userAddress: string,
  userOption: BigNumber,
  totalStaked: Proposal['totalStaked'],
  daoBounty: BigNumber
) => {
  if (userOption === undefined || userOption === null) return BigNumber.from(0);
  const userStakedOptionIndex: number = userOption.toNumber() - 1;
  let userStakedOptionString: string;
  if (userStakedOptionIndex === 0) userStakedOptionString = 'against';
  if (userStakedOptionIndex === 1) userStakedOptionString = 'for';
  if (!userStakedOptionString)
    throw new Error('Invalid userStakedOptionString');

  const stakedInUserVote: BigNumber =
    totalStaked[userStakedOptionIndex].add(userCurrentStake);
  const stakedInUserVoteFixed: FixedNumber =
    FixedNumber.fromValue(stakedInUserVote);

  if (stakedInUserVote.isZero()) return BigNumber.from(0);

  const totalStakedWithoutDaoBounty = userCurrentStake
    .add(totalStaked[0])
    .add(totalStaked[1])
    .sub(daoBounty);

  // Calculate what the user staked
  const previouslyStakedByUser = stakes[userStakedOptionString].reduce(
    (acc: BigNumber, cur) => {
      if (cur.staker.toLowerCase() === userAddress.toLowerCase()) {
        return acc.add(cur.amount);
      }
      return acc;
    },
    BigNumber.from(0)
  );

  const totalUserStake = previouslyStakedByUser.add(userCurrentStake);

  // Calculate the coefficient for the base reward

  const initialRewardCoefficient = FixedNumber.fromValue(
    totalStakedWithoutDaoBounty
  ).divUnsafe(stakedInUserVoteFixed);

  const initialReward = FixedNumber.fromValue(totalUserStake).mulUnsafe(
    initialRewardCoefficient
  );

  const initialRewardBN = BigNumber.from(
    initialReward.toString().split('.')[0]
  );

  // If the user voted for "NO", then the reward is just the initial reward
  if (userOption.eq(1)) return initialRewardBN;

  // If the user voted for "YES", then the reward is the initial reward + a percentage of the DAO bounty
  const daoBountyRewardCoefficient = FixedNumber.fromValue(daoBounty).divUnsafe(
    stakedInUserVoteFixed
  );

  const daoBountyReward = FixedNumber.fromValue(totalUserStake).mulUnsafe(
    daoBountyRewardCoefficient
  );

  const daoBountyRewardBN = BigNumber.from(
    daoBountyReward.toString().split('.')[0]
  );

  return initialRewardBN.add(daoBountyRewardBN);
};
