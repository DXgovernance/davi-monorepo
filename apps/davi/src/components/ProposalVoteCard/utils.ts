import { VoteOnProposalProps } from './types';

export const checkVotingPower = ({
  votingPowerAtProposalSnapshot,
  votingPowerAtProposalCurrentSnapshot,
}: VoteOnProposalProps) => {
  const hasNoVotingPower =
    votingPowerAtProposalSnapshot &&
    Number(votingPowerAtProposalSnapshot?.toString()) <= 0;

  const hasVotingPowerAtCurrentSnapshot =
    votingPowerAtProposalCurrentSnapshot &&
    Number(votingPowerAtProposalCurrentSnapshot?.toString()) > 0;
  return {
    hasNoVotingPower,
    hasVotingPowerAtCurrentSnapshot,
  };
};
