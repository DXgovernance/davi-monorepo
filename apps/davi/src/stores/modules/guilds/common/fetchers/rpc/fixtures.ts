export const MOCK_CONTRACT_ADDRESS =
  '0x0000000000000000000000000000000000000001';
export const MOCK_GUILD_ADDRESS = '0x0000000000000000000000000000000000000002';
