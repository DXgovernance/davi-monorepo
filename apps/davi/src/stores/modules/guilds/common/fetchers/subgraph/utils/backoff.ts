import { useBlockNumber } from 'wagmi';

export const useBackoff = (refetch: () => any) => {
  const { data: block } = useBlockNumber({
    watch: true,
  });
  let currentBlock = 0;

  // Backoff and retry logic
  const backoff = (fun, successFun, failureFun, exponent) => {
    setTimeout(async () => {
      const { data } = await fun();
      if (currentBlock < 1) {
        console.log('Still fetching current block');
      } else if (data?._meta.block.number > currentBlock) {
        successFun();
      } else if (exponent <= 20) {
        backoff(fun, successFun, failureFun, exponent + 1);
      } else {
        failureFun();
      }
    }, Math.pow(2, exponent) + Math.random() * 10000);
  };

  return async () => {
    currentBlock = block;
    backoff(
      refetch,
      () => (currentBlock = 0),
      () => console.log('Failed to fetch new data'),
      0
    );
  };
};
