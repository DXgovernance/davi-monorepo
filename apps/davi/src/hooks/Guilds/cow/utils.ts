import { GNOSIS_ID, LOCALHOST_ID, MAINNET_ID } from 'utils';

export const isAvailableOnCow = (chainId: number) => {
  return [MAINNET_ID, GNOSIS_ID, LOCALHOST_ID].includes(chainId);
};
