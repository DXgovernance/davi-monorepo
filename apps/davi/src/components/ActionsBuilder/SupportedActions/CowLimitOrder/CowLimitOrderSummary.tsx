import { useCallback, useEffect, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
  AmountContainer,
  DetailBody,
  DetailHeader,
  OrderViewContainer,
  SummaryRow,
} from './CowLimitOrderSummary.styled';
import { SummaryProps } from './types';
import { BlockExplorerLink } from 'components/primitives/Links';
import { SupportedAction } from 'components/ActionsBuilder/types';
import { Segment } from '../common/infoLine';
import { FiArrowRight } from 'react-icons/fi';
import useBigNumberToNumber from 'hooks/Guilds/conversions/useBigNumberToNumber';
import { BigNumber } from 'ethers';
import { useNetwork } from 'wagmi';
import { useTokenList } from 'hooks/Guilds/tokens/useTokenList';
import { CowQuote, useCow } from 'hooks/Guilds/cow/useCow';
import moment from 'moment';
import useTimeDifferenceHumanized from 'hooks/Guilds/time/useTimeDifferenceHumanized';

const CowLimitOrderSummary = ({ decodedCall }: SummaryProps) => {
  const { t } = useTranslation();

  const { chain } = useNetwork();
  const { tokens } = useTokenList(chain?.id, false);
  const [order, setOrder] = useState<CowQuote>(null);

  const findTokenByAddress = useCallback(
    (address: string) => {
      return tokens.find(
        token => token.address?.toLowerCase() === address?.toLowerCase()
      );
    },
    [tokens]
  );

  const { getOrder } = useCow();

  const parsedData = useMemo(() => {
    if (!decodedCall) return null;
    return {
      orderId: decodedCall.args.orderUid,
    };
  }, [decodedCall]);

  useEffect(() => {
    if (!parsedData?.orderId) return;

    const fetchOrder = async () => {
      try {
        const order = await getOrder(parsedData?.orderId);
        setOrder(order);
      } catch (error) {
        console.log('error retrieving order', error);
      }
    };

    fetchOrder();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [parsedData?.orderId]);

  const roundedBuyBalance = useBigNumberToNumber(
    order?.buyAmount ? BigNumber.from(order?.buyAmount) : null,
    order?.buyToken ? findTokenByAddress(order?.buyToken).decimals : null,
    6
  );

  const roundedSellBalance = useBigNumberToNumber(
    order?.sellAmount ? BigNumber.from(order?.sellAmount) : null,
    order?.sellToken ? findTokenByAddress(order?.sellToken)?.decimals : null,
    6
  );

  const roundedFee = useBigNumberToNumber(
    order?.feeAmount ? BigNumber.from(order?.feeAmount) : null,
    order?.sellToken ? findTokenByAddress(order?.sellToken)?.decimals : null,
    6
  );

  const isLimitOrder = useMemo(() => {
    return decodedCall?.callType === SupportedAction.COW_SWAP_LIMIT_ORDER;
  }, [decodedCall?.callType]);

  const validToMoment = moment.unix(order?.validTo);

  const expTimeHumanized = useTimeDifferenceHumanized(validToMoment);
  /**
   * With the existing action builder architecture, the limit order approval also uses this summary view
   */
  if (!isLimitOrder) {
    return null;
  }

  return (
    <>
      <DetailHeader>
        {`${t('actionBuilder.action.executeLimitOrder')}`}
      </DetailHeader>

      <OrderViewContainer>
        <SummaryRow>
          <Segment>
            <AmountContainer>{roundedSellBalance}</AmountContainer>
            <BlockExplorerLink address={order?.sellToken} showAvatar />
          </Segment>
          <Segment>
            <FiArrowRight />
          </Segment>
          <Segment>
            <AmountContainer>{roundedBuyBalance}</AmountContainer>
            <BlockExplorerLink address={order?.buyToken} showAvatar />
          </Segment>
        </SummaryRow>
        <SummaryRow>Fee: {roundedFee}</SummaryRow>
        <SummaryRow>Expires in: {expTimeHumanized}</SummaryRow>
      </OrderViewContainer>

      <OrderViewContainer>
        <DetailBody>
          <BlockExplorerLink address={decodedCall?.to} showAvatar />
        </DetailBody>
      </OrderViewContainer>
    </>
  );
};

export default CowLimitOrderSummary;
