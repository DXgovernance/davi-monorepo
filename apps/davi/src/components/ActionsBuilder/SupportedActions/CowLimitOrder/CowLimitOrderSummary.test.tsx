import { BigNumber } from 'ethers';
import { render } from 'utils/tests';
import { mockChain } from 'components/Web3Modals/fixtures';
import { limitOrderDecodedCallMock } from './fixtures';
import CowLimitOrderSummary from './CowLimitOrderSummary';

const mockBigNumber = BigNumber.from(100000000);

jest.mock('hooks/Guilds/erc20/useERC20Info', () => ({
  useERC20Info: () => ({
    name: 'Test ERC20',
    symbol: 'TEST',
    decimals: 18,
    totalSupply: mockBigNumber,
  }),
}));

jest.mock('wagmi', () => ({
  useContractRead: () => ({ data: '' }),
  useNetwork: () => ({ chain: mockChain, chains: [mockChain] }),
  useEnsResolver: () => ({
    data: {
      name: 'name.eth',
      address: '0x0000000000000000000000000000000000000000',
      contentHash: '0x0',
    },
  }),
  useEnsName: () => ({
    data: 'name.eth',
  }),
  useEnsAddress: () => ({
    data: '0x0000000000000000000000000000000000000000',
  }),
  useContractReads: () => ({
    data: [{}],
  }),
}));

jest.mock('provider', () => ({
  getBlockExplorerUrl: () => null,
}));

describe('CowLimitOrderSummary', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <CowLimitOrderSummary decodedCall={limitOrderDecodedCallMock} />
    );
    expect(container).toMatchSnapshot();
  });
  it('Should match snapshot without data', () => {
    const { container } = render(<CowLimitOrderSummary decodedCall={null} />);
    expect(container).toMatchSnapshot();
  });

  it('Should match snapshot with default values', () => {
    const { container } = render(
      <CowLimitOrderSummary decodedCall={limitOrderDecodedCallMock} />
    );
    expect(container).toMatchSnapshot();
  });
});
