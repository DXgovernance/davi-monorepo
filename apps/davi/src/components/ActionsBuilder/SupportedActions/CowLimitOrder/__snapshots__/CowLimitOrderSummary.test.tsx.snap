// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CowLimitOrderSummary Should match snapshot 1`] = `
.c0 {
  box-sizing: 'border-box';
  min-width: 0;
  margin: 0;
  padding: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: 0.4rem;
  margin-bottom: 1.5rem;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c2 {
  color: #fff;
  margin-top: 0;
  white-space: pre-wrap;
}

.c6 {
  color: #A1A6B0;
  margin: 0;
}

.c5 {
  margin-right: 1rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0.4rem 0;
}

.c4 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  display: Flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  border-radius: 10px;
  margin: 0;
  padding: 0;
}

.c8 {
  color: #A1A6B0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  overflow-wrap: break-word;
}

.c8:hover {
  cursor: pointer;
  color: #fff;
}

.c9 {
  overflow-wrap: break-word;
}

<div>
   
  <div
    class="c0 c1 c2"
  >
    actionBuilder.action.executeLimitOrder
  </div>
  <div
    class="c0 c1"
  >
    <div
      class="c0 c3"
    >
      <span
        class="c4"
      >
        <span
          class="c5"
        />
      </span>
      <span
        class="c4"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
          <polyline
            points="12 5 19 12 12 19"
          />
        </svg>
      </span>
      <span
        class="c4"
      >
        <span
          class="c5"
        />
      </span>
    </div>
    <div
      class="c0 c3"
    >
      Fee: 
    </div>
    <div
      class="c0 c3"
    >
      Expires in: 
      NaN duration.years_other NaN duration.months_other
    </div>
  </div>
  <div
    class="c0 c1"
  >
    <div
      class="c0 c1 c6"
    >
      <div
        class="c7"
        data-testid="block-explorer-container"
        direction="row"
      >
        <span
          class="c4"
        >
          <div
            class="paper"
            style="border-radius: 50px; display: inline-block; margin: 0px; overflow: hidden; padding: 0px; background-color: rgb(245, 163, 0); height: 24px; width: 24px;"
          >
            <svg
              height="24"
              width="24"
              x="0"
              xmlns="http://www.w3.org/2000/svg"
              y="0"
            >
              <rect
                fill="#18B7F2"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-3.64229549197103 3.9302875273080358) rotate(292.6 12 12)"
                width="24"
                x="0"
                y="0"
              />
              <rect
                fill="#F27E02"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-10.586767133036869 10.774685878518335) rotate(149.2 12 12)"
                width="24"
                x="0"
                y="0"
              />
              <rect
                fill="#FA5300"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-17.146496219352098 3.405876630260933) rotate(176.4 12 12)"
                width="24"
                x="0"
                y="0"
              />
              <rect
                fill="#018E80"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-24.089760410360498 10.109035008456289) rotate(161.5 12 12)"
                width="24"
                x="0"
                y="0"
              />
            </svg>
          </div>
        </span>
        <a
          class="c8"
          rel="noopener"
          target="_blank"
        >
          <div
            class="c0 c9"
          >
            name.eth
          </div>
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
            />
            <polyline
              points="15 3 21 3 21 9"
            />
            <line
              x1="10"
              x2="21"
              y1="14"
              y2="3"
            />
          </svg>
        </a>
      </div>
    </div>
  </div>
   
</div>
`;

exports[`CowLimitOrderSummary Should match snapshot with default values 1`] = `
.c0 {
  box-sizing: 'border-box';
  min-width: 0;
  margin: 0;
  padding: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: 0.4rem;
  margin-bottom: 1.5rem;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c2 {
  color: #fff;
  margin-top: 0;
  white-space: pre-wrap;
}

.c6 {
  color: #A1A6B0;
  margin: 0;
}

.c5 {
  margin-right: 1rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0.4rem 0;
}

.c4 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  display: Flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  border-radius: 10px;
  margin: 0;
  padding: 0;
}

.c8 {
  color: #A1A6B0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  overflow-wrap: break-word;
}

.c8:hover {
  cursor: pointer;
  color: #fff;
}

.c9 {
  overflow-wrap: break-word;
}

<div>
   
  <div
    class="c0 c1 c2"
  >
    actionBuilder.action.executeLimitOrder
  </div>
  <div
    class="c0 c1"
  >
    <div
      class="c0 c3"
    >
      <span
        class="c4"
      >
        <span
          class="c5"
        />
      </span>
      <span
        class="c4"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
          <polyline
            points="12 5 19 12 12 19"
          />
        </svg>
      </span>
      <span
        class="c4"
      >
        <span
          class="c5"
        />
      </span>
    </div>
    <div
      class="c0 c3"
    >
      Fee: 
    </div>
    <div
      class="c0 c3"
    >
      Expires in: 
      NaN duration.years_other NaN duration.months_other
    </div>
  </div>
  <div
    class="c0 c1"
  >
    <div
      class="c0 c1 c6"
    >
      <div
        class="c7"
        data-testid="block-explorer-container"
        direction="row"
      >
        <span
          class="c4"
        >
          <div
            class="paper"
            style="border-radius: 50px; display: inline-block; margin: 0px; overflow: hidden; padding: 0px; background-color: rgb(245, 163, 0); height: 24px; width: 24px;"
          >
            <svg
              height="24"
              width="24"
              x="0"
              xmlns="http://www.w3.org/2000/svg"
              y="0"
            >
              <rect
                fill="#18B7F2"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-3.64229549197103 3.9302875273080358) rotate(292.6 12 12)"
                width="24"
                x="0"
                y="0"
              />
              <rect
                fill="#F27E02"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-10.586767133036869 10.774685878518335) rotate(149.2 12 12)"
                width="24"
                x="0"
                y="0"
              />
              <rect
                fill="#FA5300"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-17.146496219352098 3.405876630260933) rotate(176.4 12 12)"
                width="24"
                x="0"
                y="0"
              />
              <rect
                fill="#018E80"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-24.089760410360498 10.109035008456289) rotate(161.5 12 12)"
                width="24"
                x="0"
                y="0"
              />
            </svg>
          </div>
        </span>
        <a
          class="c8"
          rel="noopener"
          target="_blank"
        >
          <div
            class="c0 c9"
          >
            name.eth
          </div>
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
            />
            <polyline
              points="15 3 21 3 21 9"
            />
            <line
              x1="10"
              x2="21"
              y1="14"
              y2="3"
            />
          </svg>
        </a>
      </div>
    </div>
  </div>
   
</div>
`;

exports[`CowLimitOrderSummary Should match snapshot without data 1`] = `
<div>
   
   
</div>
`;
